fsspec/dircache.py

```diff
@@ -0,0 +1,50 @@
+from functools import lru_cache
+import time
+
+
+class DirCache(dict):
+    """
+    Caching of directory listings, in a structure like
+
+    {"path0": [
+        {"name": "path0/file0",
+         "size": 123,
+         "type": "file",
+         ...
+        },
+        {"name": "path0/file1",
+        },
+        ...
+        ],
+     "path1": [...]
+    }
+
+    Parameters to this class control listing expiry or indeed turn
+    caching off
+    """
+    def __init__(self, usecache=True, expiry_time=False, max_size=None):
+        self._cache = {}
+        self._times = {}
+        if max_size:
+            self.q = lru_cache(max_size + 1)(lambda key: self._cache.pop(key, None))
```

Comment:
It feels a bit strange to have the max_size parameter affect the attribute where we actually store things. That leads to stuff like `if self.max` later on. Can we just always store it in the same place?

Reply:
`._cache` stored the listings. The q and times dicts hold only information on what to automatically evict, and are not accessed if eviction is not desired.